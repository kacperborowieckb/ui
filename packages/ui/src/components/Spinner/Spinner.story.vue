<template>
  <Story>
    <Variant title="Spinner">
      <div class="flex gap-4">
        <Button>
          <Spinner />
        </Button>
        <Spinner variant="background" />
      </div>
    </Variant>
  </Story>
</template>

<script setup lang="tsx">
import Button from '../Button/Button.vue'
import Spinner from './Spinner.vue'
</script>
