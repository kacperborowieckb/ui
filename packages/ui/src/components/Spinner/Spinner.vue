<template>
  <span class="relative flex size-5 items-center justify-center" aria-label="Loading">
    <span
      v-for="(spinnerLine, index) in spinnerLines" :key="index"
      class="spinner-line transform-[rotate(var(--line-rotation))_translate(150%)] absolute block h-[12%] w-1/4 rounded-[1px]"
      :class="spinnerColorClasses"
      v-bind="spinnerLine"
    />
  </span>
</template>

<script setup lang="ts">
import { computed } from 'vue'

interface SpinnerProps {
  variant: 'foreground' | 'background'
}

const props = withDefaults(defineProps<SpinnerProps>(), {
  variant: 'foreground',
})

const spinnerColorClasses = computed(() => props.variant === 'foreground' ? ' bg-primary-foreground' : 'bg-foreground')

const spinnerLines = Array.from({ length: 12 }, (_, i) => {
  return { style: { '--line-rotation': `${(30 * i) || 0.00001}deg`, '--line-delay': `${((12 - i) * -0.1).toFixed(2)}s` } }
})
</script>

<style scoped lang="css">
.spinner-line {
  animation: spin-steps 1.2s infinite;
  animation-delay: var(--line-delay);
}

@keyframes spin-steps {
  from {
    opacity: 0.8;
  } to {
    opacity: 0.2;
  }
}
</style>
