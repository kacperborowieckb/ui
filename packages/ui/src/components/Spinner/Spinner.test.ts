import { shallowMount } from '@vue/test-utils'

import Spinner from './Spinner.vue'

describe('spinner', () => {
  it('should render correctly', () => {
    const wrapper = shallowMount(Spinner)

    expect(wrapper.html()).toMatchSnapshot()
  })
})
