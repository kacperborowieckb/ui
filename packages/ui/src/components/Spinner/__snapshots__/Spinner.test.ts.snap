// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`spinner > should render correctly 1`] = `"<span data-v-bc6d31ca="" class="relative flex size-5 items-center justify-center" aria-label="Loading"><span data-v-bc6d31ca="" class="spinner-line transform-[rotate(var(--line-rotation))_translate(125%)] absolute block h-[12%] w-1/3 rounded-[1px] bg-primary-foreground" style="--line-rotation: 0.00001deg; --line-delay: -1.00s;"></span><span data-v-bc6d31ca="" class="spinner-line transform-[rotate(var(--line-rotation))_translate(125%)] absolute block h-[12%] w-1/3 rounded-[1px] bg-primary-foreground" style="--line-rotation: 36deg; --line-delay: -0.90s;"></span><span data-v-bc6d31ca="" class="spinner-line transform-[rotate(var(--line-rotation))_translate(125%)] absolute block h-[12%] w-1/3 rounded-[1px] bg-primary-foreground" style="--line-rotation: 72deg; --line-delay: -0.80s;"></span><span data-v-bc6d31ca="" class="spinner-line transform-[rotate(var(--line-rotation))_translate(125%)] absolute block h-[12%] w-1/3 rounded-[1px] bg-primary-foreground" style="--line-rotation: 108deg; --line-delay: -0.70s;"></span><span data-v-bc6d31ca="" class="spinner-line transform-[rotate(var(--line-rotation))_translate(125%)] absolute block h-[12%] w-1/3 rounded-[1px] bg-primary-foreground" style="--line-rotation: 144deg; --line-delay: -0.60s;"></span><span data-v-bc6d31ca="" class="spinner-line transform-[rotate(var(--line-rotation))_translate(125%)] absolute block h-[12%] w-1/3 rounded-[1px] bg-primary-foreground" style="--line-rotation: 180deg; --line-delay: -0.50s;"></span><span data-v-bc6d31ca="" class="spinner-line transform-[rotate(var(--line-rotation))_translate(125%)] absolute block h-[12%] w-1/3 rounded-[1px] bg-primary-foreground" style="--line-rotation: 216deg; --line-delay: -0.40s;"></span><span data-v-bc6d31ca="" class="spinner-line transform-[rotate(var(--line-rotation))_translate(125%)] absolute block h-[12%] w-1/3 rounded-[1px] bg-primary-foreground" style="--line-rotation: 252deg; --line-delay: -0.30s;"></span><span data-v-bc6d31ca="" class="spinner-line transform-[rotate(var(--line-rotation))_translate(125%)] absolute block h-[12%] w-1/3 rounded-[1px] bg-primary-foreground" style="--line-rotation: 288deg; --line-delay: -0.20s;"></span><span data-v-bc6d31ca="" class="spinner-line transform-[rotate(var(--line-rotation))_translate(125%)] absolute block h-[12%] w-1/3 rounded-[1px] bg-primary-foreground" style="--line-rotation: 324deg; --line-delay: -0.10s;"></span></span>"`;
